Src/HALALMock/Services/Encoder/Encoder.cpp

```diff
@@ -22,128 +22,73 @@ uint8_t Encoder::inscribe(Pin& pin1, Pin& pin2) {
         return;
     }
-
-    TimerPeripheral* timer = pin_timer_map[registered_encoder[id]];
-
-    if (HAL_TIM_Encoder_GetState(timer->handle) == HAL_TIM_STATE_RESET) {
-        ErrorHandler("Unable to get state from encoder in timer %s",
-                     timer->name.c_str());
-        return;
-    }
-
-    if (HAL_TIM_Encoder_Start(timer->handle, TIM_CHANNEL_ALL) != HAL_OK) {
-        ErrorHandler("Unable to start encoder in timer %s",
-                     timer->name.c_str());
-        return;
-    }
-
-    reset(id);
+    std::pair<Pin,Pin> pair_pin = Encoder::registered_encoder[id];
+    EmulatedPin &pin1_data = SharedMemory::get_pin(pair_pin.first);
+    pin1_data.PinData.ENCODER.is_on = true;
+    
 }
 
 void Encoder::turn_off(uint8_t id) {
     if (not Encoder::registered_encoder.contains(id)) {
         ErrorHandler("No encoder registered with id %u", id);
         return;
     }
-
-    TimerPeripheral* timer = pin_timer_map[registered_encoder[id]];
-
-    if (HAL_TIM_Encoder_Stop(timer->handle, TIM_CHANNEL_ALL) != HAL_OK) {
-        ErrorHandler("Unable to stop encoder in timer %s", timer->name.c_str());
-    }
+    std::pair<Pin,Pin> pair_pin = Encoder::registered_encoder[id];
+    EmulatedPin &pin1_data = SharedMemory::get_pin(pair_pin.first);
+    pin1_data.PinData.ENCODER.is_on = false;
 }
 
-void Encoder::reset(uint8_t id) {
+void Encoder::reset(uint8_t id) {}
+
+uint32_t Encoder::get_counter(uint8_t id) {
     if (not Encoder::registered_encoder.contains(id)) {
         ErrorHandler("No encoder registered with id %u", id);
         return;
     }
-
-    TimerPeripheral* timer = pin_timer_map[registered_encoder[id]];
-
-    timer->handle->Instance->CNT = UINT32_MAX / 2;
-}
-
-uint32_t Encoder::get_counter(uint8_t id) {
-    // if (not Encoder::registered_encoder.contains(id)) {
-    // 	ErrorHandler("No encoder registered with id %u", id);
-    // 	return 0;
-    // }
-
-    TimerPeripheral* timer = pin_timer_map[registered_encoder[id]];
-
-    return timer->handle->Instance->CNT;
+    std::pair<Pin,Pin> pair_pin = Encoder::registered_encoder[id];
+    EmulatedPin &pin1_data = SharedMemory::get_pin(pair_pin.first);
+    return pin1_data.PinData.ENCODER.count_value;
 }
 
 bool Encoder::get_direction(uint8_t id) {
     if (not Encoder::registered_encoder.contains(id)) {
         ErrorHandler("No encoder registered with id %u", id);
         return false;
     }
-
-    TimerPeripheral* timer = pin_timer_map[registered_encoder[id]];
-
-    return ((timer->handle->Instance->CR1 & 0b10000) >> 4);
+    std::pair<Pin,Pin> pair_pin = Encoder::registered_encoder[id];
+    EmulatedPin &pin1_data = SharedMemory::get_pin(pair_pin.first);
+    return pin1_data.PinData.ENCODER.direction;
 }
 
 uint32_t Encoder::get_initial_counter_value(uint8_t id) {
-    TimerPeripheral* timer = pin_timer_map[registered_encoder[id]];
-    return (timer->handle->Instance->ARR / 2);
+    return 0;
```

Comment:
By looking at the original code this should return exactly the half point of the range of the variable you are using to store the counter (e.g. for uint32_t -> UINT32_MAX / 2)

Reply:
Yes is a good idea